## Running all Scripts
The core functions are implemented in the kim_masterlist module, which can be run with the following command from the
 root of the repository:
```commandline
(test_env) C:\\...\\VHL-Scripts>py -m kim_masterlist
```

Without any commandline arguments, the scripts will run to their entirety from the online masterlist, running 
clustering, validation, and producing all output files and figures.


## Package Structure
The Kim masterlist, stored on Google drive at the time of writing, is a multi-sheet storage of all Kim students' 
extracted germline variants from papers published before November 2019. The code related to accessing and downloading 
these variants is located in the directory:
```commandline
kim_masterlist\\fetching
```

Code related to preprocessing, summarizing, clustering, running statistical tests, and viewing of the variants in 
the masterlist are located in the directory:
```commandline
kim_masterlist\\features
```

### Files
Any files required by or produced by the scripts are found in the directory:
```commandline
kim_masterlist\\files
```

#### Input
The input files are automatically generated through the fetching scripts, and thus do not need to be put there manually. 
Input is composed of 9 csv files, each one representing one of the students' extracted variants

#### Library
The lib directory contains data from external sources:
```commandline
kim_masterlist\\files\\lib\\Homo_sapiens_VHL_201_sequence.fa - the VHL fasta sequence used in analysis
kim_masterlist\\files\\lib\\hp.obo.txt - the human phenotype ontology used for phenotype standardization
kim_masterlist\\files\\lib\\so.obo.txt - the sequence ontology, used for variation standardization
```

#### Output
All output files are generated by running the entire package without any commandline arguments. Output is split into
multiple directories. The following are the high-level directories for kindred, patient, and variant -based analysis:
```commandline
kim_masterlist\\files\\output\\kindred
kim_masterlist\\files\\output\\patient 
kim_masterlist\\files\\output\\variant
```

Each of these directories are further divided into cluster data, summary data, summary figures, and statistical tests
for each of the analysis types:
```commandline
kim_masterlist\\files\\output\\{analysis type}\cluster
kim_masterlist\\files\\output\\{analysis type}\data
kim_masterlist\\files\\output\\{analysis type}\figures
kim_masterlist\\files\\output\\{analysis type}\tests
```

The output directory also contains high-level summary data, as well as cross-validation data for UMD and LitVar, found
in the directories:
```commandline
kim_masterlist\\files\\output\\summary
kim_masterlist\\files\\output\\validation
```

## Figures
All relevant output figures are saved in the "output" folder that gets created after running the scripts. The folders
are separated by patient, kindred, and variant. Inside these separate folders are the "data" and "figures" subfolders; the "figures"
folder contains the matplotlib-generated .eps figures, and the "data" folder contains the raw data used to generate the figures.

Of interest are:

##### Fig 1. Age-related penetrance for patients that present with a single phenotype
```commandline
File: kim_masterlist\\files\\output\\patient\figures\penetrance.eps
Code: kim_masterlist\features\kimstudents_dataframe_views.py
```
##### Fig. 2. Phenotype co-occurrence ratios for (A) patient-, (B) family- and (C) variant-based data
```commandline
File: kim_masterlist\\files\\output\\(patient, family, or variant)\figures\phenotype_correlation_ratio.eps
Code: kim_masterlist\features\kimstudents_dataframe_views.py
```
##### Fig 3. Frequency of missense variants along the VHL gene for (A) patient-, (B) family- and (C) variant-based data. 
```commandline
File: kim_masterlist\\files\\output\\(patient, family, or variant)\figures\codon_histogram.eps
Code: kim_masterlist\features\kimstudents_dataframe_views.py
```
##### Fig 4. Distribution of truncating and non-truncating variant by phenotype for (A) patient-, (B) family- and (C) variant-based data
```commandline
File: kim_masterlist\\files\\output\\(patient, family, or variant)\figures\grouped_mutant_type_counts.eps
Code: kim_masterlist\features\kimstudents_dataframe_views.py
```
##### Fig 5. Frequency of coding variants in protein and functional domains for (A) patient-, (B) family- and (C) variant-based data
```commandline
Files: kim_masterlist\\files\\output\\(patient, family, or variant)\figures\regions.eps
Code: kim_masterlist\features\kimstudents_dataframe_views.py
```
##### Fig 6. Cluster phenotype, variant type, variant domain and codon distribution for two patient clusters.
```commandline
Files:  kim_masterlist\\files\\output\patient\cluster\cluster_labels_best\clustered_generalized_phenotype_counts.eps
        kim_masterlist\\files\\output\patient\cluster\cluster_labels_best\clustered_domain_counts.eps
        kim_masterlist\\files\\output\patient\cluster\cluster_labels_best\clustered_grouped_mutation_type_counts.eps
        kim_masterlist\\files\\output\patient\cluster\cluster_labels_best\clustered_codon_start.eps

Code: kim_masterlist\features\kimstudents_dataframe_views.py
``` 
##### Fig 7. Cluster phenotype (A), variant type (B), variant domain (C) and codon statistics (D) for each of the 4 patient clusters
```commandline
Files:  kim_masterlist\\files\\output\patient\cluster\cluster_labels_second\clustered_generalized_phenotype_counts.eps
        kim_masterlist\\files\\output\patient\cluster\cluster_labels_second\clustered_domain_counts.eps
        kim_masterlist\\files\\output\patient\cluster\cluster_labels_second\clustered_grouped_mutation_type_counts.eps
        kim_masterlist\\files\\output\patient\cluster\cluster_labels_second\clustered_codon_start.eps

Code: kim_masterlist\features\kimstudents_dataframe_views.py
``` 
## Supplementary Files
The scripts output the fetched individual input files to the /output directory, which get combined into a single masterlist.
The first step of filtering is to drop all entries that do not have either a phenotype or a mutation- the output of this step is
saved in:
```
kim_masterlist\files\output\summary\postdropsupplementary_1.csv
``` 
The next step is to create additional annotated columns with clean values to 
be used in further analysis. The resulting file is:
```
kim_masterlist\files\output\summary\filtered_out.csv
``` 
#### Table 1
The summary statistics for the total number of patients, kindreds, and variants are saved to:
```
kim_masterlist\files\output\summary\postdropsummary.csv
``` 
The breakdown of these summaries by phenotype and mutation type are saved in :
```
kim_masterlist\files\output\summary\summary_by_type.csv
```
Together, these two files are used to create Table 1.